src/v/datalake/proto_to_arrow_struct.cc

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+#include "datalake/proto_to_arrow_struct.h"
+
+#include <fmt/format.h>
+
+arrow::Status datalake::detail::proto_to_arrow_struct::finish_batch() {
+    if (!_arrow_status.ok()) {
+        return _arrow_status;
+    }
+    arrow::Result<std::shared_ptr<arrow::Array>> builder_result
+      = _builder->Finish();
+
+    _arrow_status = builder_result.status();
+    std::shared_ptr<arrow::Array> array;
+    if (!_arrow_status.ok()) {
+        return _arrow_status;
+    }
+
+    // Safe because we validated the status after calling `Finish`
+    _values.push_back(std::move(builder_result).ValueUnsafe());
+
+    return _arrow_status;
+}
+arrow::Status datalake::detail::proto_to_arrow_struct::add_top_level_message(
+  const google::protobuf::Message* msg) {
+    if (!_arrow_status.ok()) {
+        return _arrow_status;
+    }
+    for (size_t child_idx = 0; child_idx < _child_arrays.size(); child_idx++) {
+        _arrow_status = _child_arrays[child_idx]->add_value(
+          msg, int(child_idx));
+        if (!_arrow_status.ok()) {
+            return _arrow_status;
+        }
+    }
+    _arrow_status = _builder->Append();
+
+    return _arrow_status;
+}
+arrow::Status datalake::detail::proto_to_arrow_struct::add_value(
+  const google::protobuf::Message* msg, int field_idx) {
+    if (!_arrow_status.ok()) {
+        return _arrow_status;
+    }
+    auto desc = msg->GetDescriptor()->field(field_idx);
```

Comment:
Do we need to validate that desc isn't nullptr or something? What guarantees us that the descriptor has the correct fields?